ServantPersistent/src/Types.hs

```diff
@@ -0,0 +1,124 @@
+{-# LANGUAGE DeriveAnyClass  #-}
+{-# LANGUAGE DeriveFunctor  #-}
+{-# LANGUAGE TypeInType  #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE PolyKinds  #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE GADTs  #-}
+module Types
+    where
+
+import Data.Aeson
+import Data.Aeson.TH
+import Network.Wai
+import Network.Wai.Handler.Warp
+import Servant
+import Servant.API.Experimental.Auth
+import Servant.Server.Experimental.Auth.Cookie
+import Data.ByteString
+import Data.Proxy
+import GHC.Generics
+import Control.Monad.Except
+import Control.Monad.Reader
+import Data.Serialize
+import Data.Text
+import Database.Persist
+import Database.Persist.Postgresql
+import Data.Time.Clock
+import Control.Lens hiding ((.=))
+import Database.Persist.Sql
+import Database.Persist.TH
+import Data.Type.Equality
+import Data.HashMap.Strict (unions,empty)
+
+data Environment = Test | Devel | Production deriving (Eq, Show)
+
+data CookieError = NotPresent | AuthError AuthCookieException deriving (Eq, Show)
+
+data Config = Config
+    { authSettings :: AuthCookieSettings
+    , randomSource :: RandomSource
+    , serverKey :: ServerKey
+    , environment :: Environment
+    , dbPool :: ConnectionPool
+    }
+
+type App = (ReaderT Config (ExceptT ServantErr IO))
+
+data LoginForm = Login { username :: String
+                       , password :: String
+                       } deriving (Show, Generic, Serialize, FromJSON, ToJSON)
+
+data Session = Session { sessionUser :: String
+                       } deriving (Show, Generic, Serialize, FromJSON , ToJSON)
+
+class Reify s a | a -> s where
+    reify :: Proxy a -> s a
+
+data UserStatus = BlockedU | InactiveU | ActiveU
+                        deriving (Read, Show, Generic, Serialize, FromJSON , ToJSON)
+derivePersistField "UserStatus"
+
+data TenantStatus = NewT | InactiveT | ActiveT
+                        deriving (Eq, Read, Show, Generic, Serialize, FromJSON , ToJSON)
+derivePersistField "TenantStatus"
+data TSSingleton (s :: TenantStatus) where
+    TNew :: TSSingleton NewT
+    TInactive :: TSSingleton InactiveT
+    TActive :: TSSingleton ActiveT
+instance Reify TSSingleton 'NewT where
+    reify _ = TNew
+instance Reify TSSingleton 'InactiveT where
+    reify _ = TInactive
+instance Reify TSSingleton 'ActiveT where
+    reify _ = TActive
+
+class HasTimestamp s where
+    createdAt :: Lens' s UTCTime
+    updatedAt :: Lens' s UTCTime
+
+data TenantIdent =
+    TI { _name :: Text
+       , _backofficeDomain :: Text
+       } deriving (Generic)
+instance FromJSON TenantIdent where
+    parseJSON = genericParseJSON (defaultOptions { fieldLabelModifier = Prelude.drop 1})
+instance ToJSON TenantIdent where
+    toEncoding = genericToEncoding (defaultOptions { fieldLabelModifier = Prelude.drop 1})
+    toJSON = genericToJSON (defaultOptions { fieldLabelModifier = Prelude.drop 1})
+makeClassy ''TenantIdent
+type TenantInput = TenantIdent
+
+data TenantBase a =
```

Comment:
My reasoning behind the Tenant types is this:

TenantIdent contains the fields which are always required in anything to do with tenants.

TenantBase contains a TenantIdent, timestamps(which are required most of the time) and a parameter, whose type changes based on the kind of tenant we are dealing with.

`TenantBase TenantOutputField` is wrongly named. It should be something like `TenantBase OtherDBFields`

`TenantBase TenantOutputField`, whatever its called, is isomorphic to, and should always remain isomorphic to DBTenant, which is the model of the tenant in the DB. This is evidenced by `tbdbIso`, another example of my great talent for naming things.

As of now, `TenantOutput` is `TenantBase TenantOutputField`, but that can change if, for example, we need the user as well as the tenant details in the output, or some sensitive data is added to DBTenant.

As for `TenantSafe`, we pretend it does not exist. It stores the `tenantStatus` at the typelevel, but should never be explictly used in code. Instead, `Tenant` should be used whenever a Tenant with a particular status is demanded. 

I could think of two ways to implement type safe Tenants, the first using GADTs, necessitating three seperate constructors(hence interfaces) to largely the same data. I tried this out and it seemed to be too clunky to work with.

Instead, implementing Tenant as a closed, injective type family along with TenantStatus allows us to be completely safe while having only one, conveniently polymorphic constructor. According to me, this is much nicer than the GADT approach.